libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -597,6 +597,89 @@ string YulUtilFunctions::overflowCheckedIntSubFunction(IntegerType const& _type)
 	});
 }
 
+string YulUtilFunctions::overflowCheckedIntExpFunction(
+	IntegerType const& _type,
+	IntegerType const& _exponentType
+)
+{
+	solUnimplementedAssert(!_type.isSigned(), "");
+	solAssert(!_exponentType.isSigned(), "");
+	string functionName = "checked_exp_" + _type.identifier() + "_" + _exponentType.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return
+			Whiskers(R"(
+			function <functionName>(base, exponent) -> power {
+				base := <cleanupFunction>(base)
+				exponent := <exponentCleanupFunction>(exponent)
+				// 0**0 == 1
+				if iszero(exponent) { power := 1 leave }
+				if or(
+					lt(base, 2),
+					eq(exponent, 1)
+				) { power := base leave }
+
+
+				let max := <maxValue>
+
+				// small base optimizations
+				if eq(base, 2) {
+					if gt(exponent, 255) { revert(0, 0) }
+					power := exp(2, exponent)
```

Comment:
This will be converted to `shl` by the optimizer right?

Reply:
We can check, but manly, `shl` does not exist on all platforms.

Reply:
Actually it seems we do not have such a rule. I will add it separately.